WalletWasabi/Blockchain/Analysis/FeesEstimation/HybridFeeProvider.cs

```diff
@@ -0,0 +1,139 @@
+using WalletWasabi.Logging;
+using WalletWasabi.Services;
+
+namespace WalletWasabi.Blockchain.Analysis.FeesEstimation
+{
+	/// <summary>
+	/// Manages multiple fee sources. Returns the best one.
+	/// Prefers local full node, as long as the fee is accurate.
+	/// </summary>
+	public class HybridFeeProvider : IDisposable
+	{
+		private volatile bool _disposedValue = false; // To detect redundant calls
+
+		public HybridFeeProvider(WasabiSynchronizer synchronizer, RpcFeeProvider? rpcFeeProvider)
+		{
+			Synchronizer = synchronizer;
+			RpcFeeProvider = rpcFeeProvider;
+
+			Synchronizer.AllFeeEstimateArrived += OnAllFeeEstimateArrived;
+
+			if (RpcFeeProvider is not null)
+			{
+				RpcFeeProvider.AllFeeEstimateArrived += OnAllFeeEstimateArrived;
+			}
+
+			var syncerEstimate = Synchronizer.LastResponse?.AllFeeEstimate;
+			var rpcEstimate = RpcFeeProvider?.LastAllFeeEstimate;
+			var betterEstimate = rpcEstimate?.IsAccurate is true ? rpcEstimate : syncerEstimate;
+			if (betterEstimate is not null)
+			{
+				SetAllFeeEstimate(betterEstimate);
+			}
+		}
+
+		public event EventHandler<AllFeeEstimate>? AllFeeEstimateChanged;
+
+		public WasabiSynchronizer Synchronizer { get; }
+		public RpcFeeProvider? RpcFeeProvider { get; }
+		private object Lock { get; } = new object();
+		public AllFeeEstimate? AllFeeEstimate { get; private set; }
+
+		private void OnAllFeeEstimateArrived(object? sender, AllFeeEstimate fees)
+		{
+			// Only go further if we have estimations.
+			if (fees?.Estimations?.Any() is not true)
+			{
+				return;
+			}
+
+			var notify = false;
+			lock (Lock)
+			{
+				if (AllFeeEstimate is null)
+				{
+					// If the fee was never set yet, then we should set it regardless where it came from.
+					notify = SetAllFeeEstimate(fees);
+				}
+				else if (!AllFeeEstimate.IsAccurate && fees.IsAccurate)
+				{
+					// If the fee was inaccurate and the new fee is accurate, then we should set it regardless where it came from.
+					notify = SetAllFeeEstimate(fees);
+				}
+				else if (AllFeeEstimate.IsAccurate && !fees.IsAccurate)
+				{
+					// If the fee was accurate and the new fee is inaccurate, then we should leave the fees alone.
+					return;
+				}
+				else if (sender is RpcFeeProvider)
+				{
+					// If the fee is coming from the user's full node, then set it.
+					notify = SetAllFeeEstimate(fees);
+				}
+				else if (sender is WasabiSynchronizer && RpcFeeProvider is null)
+				{
+					// If fee is coming from the the backend and user doesn't use a full node, then set the fees.
+					notify = SetAllFeeEstimate(fees);
+				}
+				else if (sender is WasabiSynchronizer && RpcFeeProvider is not null && RpcFeeProvider.InError is true)
+				{
+					// If fee is coming from the the backend, user uses a full node, but it doesn't provide fees, then set the fees.
+					notify = SetAllFeeEstimate(fees);
+				}
+			}
+
+			if (notify)
+			{
+				var accuracy = fees.IsAccurate ? "Accurate" : "Inaccurate";
```

Comment:
What if we put ```fees.IsAccurate ? "Accurate" : "Inaccurate"``` inside the logging instead of creating a variable?

Reply:
You can't. For some reason string interpolation doesn't like `? :` operator.

Reply:
To make string interpolation work for `? :` you have to add it inside `( ? : )`